src/V3Width.cpp

```diff
@@ -6551,6 +6551,93 @@ class WidthVisitor final : public VNVisitor {
+            }
+            return false;
+        }
+
+        // e) Dynamic array, associative array, and queue types are equivalent if they are the same
+        // kind of array (dynamic, associative, or queue), have equivalent index types (for
+        // associative arrays), and have equivalent element types.
+        const bool lhsIsDynArray = VN_IS(lhs, DynArrayDType);
+        const bool rhsIsDynArray = VN_IS(rhs, DynArrayDType);
+        const bool lhsIsQueue = VN_IS(lhs, QueueDType);
+        const bool rhsIsQueue = VN_IS(rhs, QueueDType);
+        const bool lhsIsAssocArray = VN_IS(lhs, AssocArrayDType);
+        const bool rhsIsAssocArray = VN_IS(rhs, AssocArrayDType);
+
+        if (lhsIsDynArray || rhsIsDynArray || lhsIsQueue || rhsIsQueue || lhsIsAssocArray
+            || rhsIsAssocArray) {
+            if (VN_IS(lhs, DynArrayDType) && VN_IS(rhs, DynArrayDType)) {
+                const AstDynArrayDType* const lhsp = VN_CAST(lhs, DynArrayDType);
+                const AstDynArrayDType* const rhsp = VN_CAST(rhs, DynArrayDType);
+                return isEquivalentDType(lhsp->subDTypep(), rhsp->subDTypep());
+            }
+
+            if (VN_IS(lhs, QueueDType) && VN_IS(rhs, QueueDType)) {
+                const AstQueueDType* const lhsp = VN_CAST(lhs, QueueDType);
+                const AstQueueDType* const rhsp = VN_CAST(rhs, QueueDType);
+                return isEquivalentDType(lhsp->subDTypep(), rhsp->subDTypep());
+            }
+
+            if (VN_IS(lhs, AssocArrayDType) && VN_IS(rhs, AssocArrayDType)) {
+                const AstAssocArrayDType* const lhsp = VN_CAST(lhs, AssocArrayDType);
+                const AstAssocArrayDType* const rhsp = VN_CAST(rhs, AssocArrayDType);
+                return isEquivalentDType(lhsp->subDTypep(), rhsp->subDTypep())
+                       && isEquivalentDType(lhsp->keyDTypep(), rhsp->keyDTypep());
+            }
+            return false;
+        }
+
+        // c) Packed arrays, packed structures, packed unions, and built-in integral
+        // types are equivalent if they contain the same number of total bits, are either all
+        // 2-state or all 4-state, and are either all signed or all unsigned.
+        if (lhs->isIntegralOrPacked() && rhs->isIntegralOrPacked()) {
+            if (lhs->width() != rhs->width()) return false;
+            if (lhs->isFourstate() != rhs->isFourstate()) return false;
+            if (lhs->isSigned() != rhs->isSigned()) return false;
+            return true;
+        }
+
+        return true;
+    }
+
+    static bool isAggregateType(const AstNode* nodep) {
+        if (!nodep) return false;
+        const AstNodeDType* dtypep = nodep->dtypep();
+        if (!dtypep) return false;
+        dtypep = dtypep->skipRefp();
+        if (!dtypep) return false;
+        return VN_IS(dtypep, QueueDType) || VN_IS(dtypep, DynArrayDType)
+               || VN_IS(dtypep, UnpackArrayDType) || VN_IS(dtypep, AssocArrayDType);
```

Comment:
Probably better to make a `virtual bool AstNodeDType::isAggregateType` and overlode in these 4 classes.

```suggestion
        return dtypep->isAggregateType();
```